reader_concurrency_semaphore.cc

```diff
@@ -88,6 +89,7 @@ class reader_permit::impl
     timer<db::timeout_clock> _timeout_timer;
     abort_source _abort_source;
     query::max_result_size _max_result_size{query::result_memory_limiter::unlimited_result_size};
+    std::optional<shared_future<>> _memory_future;
 
```

Comment:
Why is this a shared_future? It can cause a thundering herd problem.

But I can see: a query completing and releasing lots of memory can indeed make room for lots of other queries.

Reply:
True. I was confused thinking it's for multiple permits (and so when we drop below the limit we release everything). It's still true, but we'll just live with it.

Do we have hysteresis? So that once we enter serialized mode, we spend some time in it to let it retire some queries and not bounce across modes.

Reply:
Not at the moment. Well actually, once we dip below the limit again, we'll admit previously serialized readers one-by-one. If the first one bumps the usage above the limit immediately once admitted, no other reads will be admitted. So we kinda do have a histeresis of sort.

Reply:
I think real hysteresis would be better, but let's cap the effort.